internal/txapp/txapp.go

```diff
@@ -492,30 +492,44 @@ func (r *TxApp) AccountInfo(ctx context.Context, acctID []byte, getUncommitted b
 	return a.Balance, a.Nonce, nil
 }
 
-func eventSize(event *types.VotableEvent) int64 {
-	return int64(len(event.Type)) + int64(len(event.Body))
-}
-
 // ProposerTxs returns the transactions that the proposer should include in the
 // next block.
 // It takes txNonce as an argument because, the proposer may have its own transactions
 // in the mempool that are included in the current block. Therefore, we need to know the
 // largest nonce of the transactions included in the block that are authored by the proposer.
 // This transaction only includes voteBodies for events whose bodies have not been received by the network.
 // Therefore, there won't be more than 1 VoteBody transaction per event.
-func (r *TxApp) ProposerTxs(ctx context.Context, txNonce uint64, maxTxSize int64) ([]*transactions.Transaction, error) {
+func (r *TxApp) ProposerTxs(ctx context.Context, txNonce uint64, maxTxSize int64, proposerAddr []byte) ([][]byte, error) {
+	bal, nonce, err := r.AccountInfo(ctx, proposerAddr, false)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get proposer account: %w", err)
+	}
+
+	if r.GasEnabled && nonce == 0 && bal.Sign() == 0 {
+		r.log.Debug("proposer account has no balance, not allowed to propose any new transactions")
+		return nil, nil
+	}
+
+	if txNonce == 0 {
+		txNonce = uint64(nonce) + 1
+	}
+
 	readTx, err := r.Database.BeginReadTx(ctx)
 	if err != nil {
 		return nil, err
 	}
 	defer readTx.Rollback(ctx) // always rollback read tx
 
 	// Consider empty vote body tx and the rlp encoding overhead size(safety buffer)
-	maxTxSize -= r.emptyVoteBodyTxSz - rlpEncodingOverheadSize
+	maxTxSize -= r.emptyVoteBodyTxSize - rlpEncodingOverheadSize
```

Comment:
Just to double check, do we want `maxTxSize -= (r.emptyVoteBodyTxSz - rlpEncodingOverheadSize)` (as you have it, equivalently) or `maxTxSize -= r.emptyVoteBodyTxSz + rlpEncodingOverheadSize` (subtract both from `maxTxSize`)?

Reply:
Oops, I wanted the second one.

Reply:
Yeah, that's probably because there is only 1 transaction that's introduced by the proposer. If we have more than one, we could account for these overhead multiple times.